Various DRM Atomic improvements

LongChair opened the pull request:
Here are a few commits that intend to improve drm atomic support.
The main topic there are taking care of are :
- switch the the new API for the drm prime interop.
- use atomic modesetting in atomic path.
- Add the ability to select planes for video & osd
- Add the ability to use a specified osd size that can be different than videomode.

cc @wm4  @xantoz

I agree that my changes can be relicensed to LGPL 2.1 or later.

wm4 commented on video/out/opengl/context_drm_egl.c:
```diff
@@ -489,12 +583,27 @@ static bool drm_egl_init(struct ra_ctx *ctx)
+        p->osd_size.width = p->kms->mode.hdisplay;
+        p->osd_size.height = p->kms->mode.vdisplay;
+    } else {
+        if (!p->kms->atomic_context) {
```
On both of these ifs, remove the `!`, and switch the if/else bodies.

xantoz commented on video/out/drm_common.c:
```diff
@@ -47,12 +47,18 @@ const struct m_sub_options drm_conf = {
                             0, drm_validate_connector_opt),
         OPT_INT("drm-mode", drm_mode_id, 0),
-        OPT_INT("drm-overlay", drm_overlay_id, 0),
+        OPT_INT("drm-osd-plane-id", drm_osd_plane_id, 0),
```
Should we not document these options as well? If we do we probably also need to clarify when what plane is use. That only when using drmprime hwdec is the video actually rendered to the video plane &c.

I guess we could consider it internal/primarily for API use due to their confusing nature...

wm4 commented on video/out/opengl/libmpv_gl.c:
```diff
@@ -73,6 +73,12 @@ static int init(struct libmpv_gpu_context *ctx, mpv_render_param *params)
+            ra_add_native_resource(ctx->ra, "drm_params", drm_params);
+        void *drm_osd_size = p->gl->MPGetNativeDisplay("drm_osd_size");
+        if (drm_osd_size)
+            ra_add_native_resource(ctx->ra, "drm_osd_size", drm_osd_size);
```
Yes.

wm4 replied:
Also not sure if I agree that the compat stuff should stay here, since the new API has proper support for it.

LongChair replied:
I understand what you mean, but unless i missed something, there is no other way to have access to this from the hwdec ... the params array is not passed there.

LongChair commented:
Fixed the comments above, thanks to lemme know if you have any further comments :)

xantoz commented on DOCS/man/vo.rst:
```diff
@@ -488,10 +488,15 @@ Available video output drivers are:
+        Index is zero based, and related to crtc.
+        (default: first overlay plane)
+
+    ``--drm-video-plane-id=<number>``
```
Sorry about the nitpicking, but we should probably mention that this is only really the case when we use drmprime hwdec and --vo=gpu --gpu-context=drm, and that otherwise (this includes --vo=drm) the video plane is unused with the osd plane instead being the "video" plane in this case.

I know it's a bit late, but I've been thinking that maybe different naming for the planes could be used here to make the matter less confusing. Maybe something like "drm-render-plane-id" (for what is now "drm-osd-plane-id") and "drm-hwdec-video-plane-id" (for what is now "drm-video-plane-id"), for example. (Although this particular naming would of course not be forward-compatible if we decide to use separate planes for video/OSD in the software decoded case in the future.) 

What do you think (and if anyone has better naming ideas please say)?

LongChair replied:
I see what you're saying. Although we have to consider that options should have a not too much technical name. hwdec probably don't mean anything for most users typically :)
But as you say, in some cases both video & osd are on the same plane (in which case btw you probably don't want any of those options, primary plane will be used then).
Maybe mentioning that those options should be used in the case you mentioned would be enough to clarify that case ?

xantoz replied:
Mentioning that the options should primarily be used together with drmprime hwdec, and that in other cases (--vo=drm, --gpu-context=drm swdec) everything will be rendered to the "osd plane" should be sufficient, yes.

Re: the names: yeah, my names were just examples. I think maybe better ones than that could be thought of.

LongChair commented:
Adjusted documentation description :)

xantoz commented on DOCS/man/vo.rst:
```diff
@@ -488,10 +488,19 @@ Available video output drivers are:
+        Select the DRM planed index to use for video layer.
+        Index is zero based, and related to crtc.
+        This option is only valid when using the drm_prime renderer which supports
+        several layers and with ``vo=gpu`` and ``gpu-context=drm``.
```
use commas or parentheses around "which supports several layers". 
"and with" -> "together with"

also perhaps "only valid" -> "only has an effect", since it is technically valid (i.e. mpv won't complain), it just won't do anything useful.

xantoz commented on DOCS/man/vo.rst:
```diff
@@ -488,10 +488,19 @@ Available video output drivers are:
+        Select the DRM planed index to use for OSD (or OSD and video).
+        Index is zero based, and related to crtc.
+        When using this option with drm_prime renderer, it will only affect
+        the OSD contents. Otherwise will set OSD & video plane.
```
"Otherwise will set..." -> "Otherwise it will set..."

LongChair commented:
grammar gripes processed :)

xantoz commented:
So I finally got around to testing it on an intel with more than one connector, and I saw some regressions (Sandybridge, linux 4.16.0, mesa-18.0.0. ). I think it's related to the new atomic modesetting primarily, although I haven't looked too much into it yet so I can't entirely rule out the new connector handling (will do later).

It might also be simply that atomic modesetting is not yet implemented for these old GPU:s, in which case we will need to have a fallback (still use atomic commits, but do the modesetting as before), however I think I've been able to run a kmscube implementation that uses atomic modesetting fine (will test later) on sandybridge.

On to the tests I did:

It works but with a warning on the internal LVDS panel, but probably only because it only supports one resolution, and it will work without modesetting: 
[lvds--working-but-probably-just-luck.log](https://github.com/mpv-player/mpv/files/1895224/lvds--working-but-probably-just-luck.log)

On the HDMI-A-1 connector it fails hard: 
[hdmi-a-1--not-working-atomic-modesetting.log](https://github.com/mpv-player/mpv/files/1895227/hdmi-a-1--not-working-atomic-modesetting.log)

If I first use the HDMI-A-1 panel using -vo drm, or just an mpv without this PR the fbcon on this VT seems to to change a bit and the LVDS screen gets deactivated (multi-screen setups is something which could use some improvement, but that is another problem for another day. both deinit and in particular VT switching are a bit flakey) and the HDMI-A-1 screen takes on a "main" role. Now I can actually do some of the lower resolutions (with warning messages) on the HDMI, but not above 1024x768 or so (I guess because of the resolution the fbcon picked for HDMI-A-1 after mpv closed). I also tried some VT switching now, to see if I would get more error messages: 
[hdmi-a-1--kinda-works--also-vtswitching.log](https://github.com/mpv-player/mpv/files/1895260/hdmi-a-1--kinda-works--also-vtswitching.log)


I guess this latter case is similar to the first with LVDS though, that we simply got lucky because the fbcon was already setup somehow.

LongChair commented:
@xantoz : any more investigations ? Would you have any suggestions regarding this PR ?
I would like to get this going :)

LongChair commented:
@wm4 : by the way, anymore comments from your side ? :)

wm4 commented on libmpv/render_gl.h:
```diff
@@ -147,6 +147,37 @@ typedef struct mpv_opengl_fbo {
+} mpv_opengl_drm_params;
+
+typedef struct mpv_opengl_drm_osd_size {
+
```
Weird newline.

wm4 commented on libmpv/render.h:
```diff
@@ -164,6 +164,18 @@ typedef enum mpv_render_param_type {
+     * Valid for mpv_render_context_create().
+     * Type : struct mpv_opengl_drm_osd_size*
+     */
+    MPV_RENDER_PARAM_DRM_OSD_SIZE = 11,
```
Apparently your code expects that the pointers the user passes are valid for the life time of the context, which I'd find surprising. I think the structs really should be copied (and then use pointers for fields which somehow need to be changed per frame or something). That would require changes in libmpv_gpu.c (init() function). I guess native_resource_map needs to be changed into an array of structs, so a struct size to copy can be specified.

LongChair replied:
The drmModeEtomicRequest pointer which is what you're referring to as changing every frame is in that struct and already passed as a pointer to the pointer address. So copying this should still be possible.
But yes you are right on the fact that I have been assuming that the struct keeps being alive during the whole playback. If some change is made to the native_resource_map, we can probably use a copy to overcome this.

wm4 replied:
Added such a change here: https://github.com/mpv-player/mpv/pull/5716/commits/0bfdb0029dd38350169677a38601a17f8c1a727a
You'd simply set the .size field to the struct sizes, and it'd copy them.

wm4 replied:
That's part of #5716 btw., which somehow turned into a PR for general changes.

xantoz commented:
@LongChair I'm still working on it, but I get problems already with commit 540fec5b6379ea739842637c0ba9dbd472843c5f (can't use --drm-format=xrgb2101010 with -vo gpu. when it does work I get black bars through it. next commit fixes this though, but introduces the modesetting errors) so it is probably not solely atomic modesetting related.

xantoz commented on video/out/drm_atomic.c:
```diff
@@ -125,16 +137,19 @@ void drm_object_print_info(struct mp_log *log, struct drm_object *object)
-struct drm_atomic_context *drm_atomic_create_context(struct mp_log *log, int fd,
-                                                     int crtc_id, int overlay_id)
+struct drm_atomic_context *drm_atomic_create_context(struct mp_log *log, int fd, int crtc_id,
+													 int connector_id, int osd_plane_id, int video_plane_id)
```
do not use tabs for indentation

xantoz commented on video/out/drm_atomic.c:
```diff
@@ -180,21 +208,27 @@ struct drm_atomic_context *drm_atomic_create_context(struct mp_log *log, int fd,
-                    }
-                    else if (value == DRM_PLANE_TYPE_PRIMARY) {
-                        ctx->primary_plane = plane;
+                } else {                    
```
trailing whitespace

xantoz commented on DOCS/man/vo.rst:
```diff
@@ -488,10 +488,19 @@ Available video output drivers are:
-        Select the DRM overlay index to use.
-        Overlay index is zero based, and related to crtc.
-        (default: 0)
+    ``--drm-osd-plane-id=<number>``
```
Move the documentation of these options from this commit, to commit 540fec5b6379ea739842637c0ba9dbd472843c5f when they are introduced in the code.
Like this the documentation will have been introduced before the options.

xantoz commented on video/out/opengl/context_drm_egl.c:
```diff
@@ -561,8 +577,9 @@ static bool drm_egl_init(struct ra_ctx *ctx)
-
+    ra_add_native_resource(ctx->ra, "drm_params", &p->drm_params);
+    ra_add_native_resource(ctx->ra, "drm_osd_size", &p->osd_size);
+    
```
trailing whitespace

xantoz commented on video/out/drm_atomic.c:
```diff
@@ -231,18 +240,34 @@ struct drm_atomic_context *drm_atomic_create_context(struct mp_log *log, int fd,
-    if (!ctx->overlay_plane) {
-        mp_err(log, "Failed to find overlay plane with id=%d\n", overlay_id);
-        goto fail;
+    // default video plane to primary if unspecified
```
default video plane to overlay

xantoz commented on video/out/drm_atomic.c:
```diff
@@ -231,18 +240,34 @@ struct drm_atomic_context *drm_atomic_create_context(struct mp_log *log, int fd,
-        mp_err(log, "Failed to find primary plane\n");
-        goto fail;
+    // default OSD plane to overlay if unspecified
+    if (!ctx->osd_plane) {
```
Does DRM guarantee that plane ID:s are never 0? If not we need to rewrite this with a separate bool flag variable.

xantoz commented on video/out/drm_atomic.c:
```diff
@@ -231,18 +240,34 @@ struct drm_atomic_context *drm_atomic_create_context(struct mp_log *log, int fd,
-        mp_err(log, "Failed to find overlay plane with id=%d\n", overlay_id);
-        goto fail;
+    // default video plane to primary if unspecified
+    if (!ctx->video_plane) {
```
same here. Does DRM guarantee that plane ID:s are never zero?

xantoz commented on video/out/drm_atomic.c:
```diff
@@ -231,18 +240,34 @@ struct drm_atomic_context *drm_atomic_create_context(struct mp_log *log, int fd,
-        goto fail;
+    // default video plane to primary if unspecified
+    if (!ctx->video_plane) {
+        if (primary_id) {
```
same here ^

xantoz commented on video/out/drm_atomic.c:
```diff
@@ -231,18 +240,34 @@ struct drm_atomic_context *drm_atomic_create_context(struct mp_log *log, int fd,
-        goto fail;
+    // default OSD plane to overlay if unspecified
+    if (!ctx->osd_plane) {
+        if (overlay_id) {
```
same here ^

xantoz commented:
You've missed to change "primary plane" to "OSD plane" (or whatever we're going to end up calling it) in a few log messages and comments in context_drm_egl.c

LongChair commented:
Updated this PR according to comments.
Also added a couple commits from @xantoz as he submitted a PR.
I also had to chnage the ` atomic_request`  member of the drm parameters struct type to a double pointer as we're now copying the render API params.

xantoz commented:
Looks generally good to me. 

I'd like to improve the options a little bit (So that we can specify --drm-osd-plane-id=primary or --drm-osd-plane-id=overlay, for getting a generic plane for instance, since using the indexes like now is hard. Maybe also adding a help option?), but I will do that in a later PR which I create, so I think we should merge this first.

LongChair commented:
@wm4 : any further comments from your side ? :)

wm4 commented:
No.

jeeb commented:
Have tested that it compiles, and that generally the code seems to look OK and documentation is added for new options. @LongChair is pretty much the DRM maintainer and this PR touches almost fully just it.
